### PR TITLE
[ci] added CODEOWNERS (fixes #2194)

### LINKED ISSUE #2194: Add CODEOWNERS
I learned from @Paxanator and @mjermann over on [fRanz](https://github.com/UptakeOpenSource/fRanz/issues/24) that GitHub supports a notion of "default reviewers" similar to [Bitbucket](https://confluence.atlassian.com/bitbucketserver/add-default-reviewers-to-pull-requests-834221295.html).

You just have to add a `.github/CODEOWNERS`, per [instructions here](https://help.github.com/en/articles/about-code-owners). You can even put owners on particular file extensions or sub-directories, which I think aligns well with [the way we've been handling it](https://github.com/microsoft/LightGBM/blob/76170788c4bd6239334408f88cf51183ec791fa2/docs/FAQ.rst#critical).

This might save us a few clicks and put into code what we've been managing by convention.

Thoughts?

### PR DESCRIPTION
**Pull request by jameslamb**
See #2194 for details. Please suggest if I've misinterpreted https://github.com/microsoft/LightGBM/blob/76170788c4bd6239334408f88cf51183ec791fa2/docs/FAQ.rst#critical...I did my best to make this file consistent with the conventions I've observed in PRs in this project. It's very likely I have made some mistakes.

My hope is that using `CODEOWNERS` will turn something being done by convention into something being done by automation. It will, I hope, also improve the contributor experience.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# This file controls default reviewers for LightGBM code.
+# See https://help.github.com/en/articles/about-code-owners
+# for details
+#
+# Maintainers are encouraged to use their best discretion in
+# setting reviewers on PRs manually, but this file should
+# offer a reasonable automatic best-guess
+
+# main C++ code
```
Comment (StrikerRUS) [inline .github/CODEOWNERS]: `src/*` ?

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+include/*    @guolinke @chivee
+CmakeLists.txt    @guolinke @chivee @Laurae2 @jameslamb @wxchan @henry0312 @StrikerRUS @huanzhang12
+
+# R code
+R-package/*    @Laurae2 @jameslamb
+*.R    @Laurae2 @jameslamb
+
+# Python code
+python-package/*    @StrikerRUS @chivee
+*.py    @StrikerRUS @wxchan @henry0312
```
Comment (StrikerRUS) [inline .github/CODEOWNERS]: This regex catches files in https://github.com/microsoft/LightGBM/tree/master/helpers folder, for example. Not sure that anyone else beside me and **guolinke**  want to maintain them...

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# main C++ code
+include/*    @guolinke @chivee
+CmakeLists.txt    @guolinke @chivee @Laurae2 @jameslamb @wxchan @henry0312 @StrikerRUS @huanzhang12
+
+# R code
+R-package/*    @Laurae2 @jameslamb
+*.R    @Laurae2 @jameslamb
+
+# Python code
+python-package/*    @StrikerRUS @chivee
```
Comment (StrikerRUS) [inline .github/CODEOWNERS]: **henry0312** should be here :)

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+*.py    @StrikerRUS @wxchan @henry0312
+
+# CI administrative stuff
+.ci/*    @StrikerRUS @Laurae2 @jameslamb
+docs/*    @StrikerRUS @Laurae2 @jameslamb
+.travis.yml    @StrikerRUS @Laurae2 @jameslamb
+.appveyor.yml    @StrikerRUS @Laurae2 @jameslamb
+.vsts-ci.yml    @Laurae2
```
Comment (StrikerRUS) [inline .github/CODEOWNERS]: Maybe merge into `*.yml` with some exceptions?

File: .github/CODEOWNERS (hunk 2 above)
Comment (jameslamb) [reply .github/CODEOWNERS]: To be honest, I never even noticed that `helpers/` directory was there!

I'll remove it and rely on `python-package/*` to catch things

File: .github/CODEOWNERS (hunk 3 above)
Comment (jameslamb) [reply .github/CODEOWNERS]: an oversight, apologies!

File: .github/CODEOWNERS (hunk 1 above)
Comment (jameslamb) [reply .github/CODEOWNERS]: yes good catch. I will also add **Laurae2** and myself for R-specific files in there

File: .github/CODEOWNERS (hunk 4 above)
Comment (jameslamb) [reply .github/CODEOWNERS]: sure!

**Comment by jameslamb** [comment]
Addressed comments! Please review at your earliest convenience.

File: .github/CODEOWNERS [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+# CI administrative stuff
+.ci/*    @StrikerRUS @Laurae2 @jameslamb
+docs/*    @StrikerRUS @Laurae2 @jameslamb
+*.yml    @StrikerRUS @Laurae2 @jameslamb
+.vsts-ci.yml    @Laurae2
+
+# GPU code
+docker/gpu/*    @huanzhang12
+docs/GPU-Windows.rst    @huanzhang12
```
Comment (StrikerRUS) [inline .github/CODEOWNERS]: Maybe `docs/GPU-*.rst`?

Comment (jameslamb) [reply .github/CODEOWNERS]: good change, yeah

**Comment by jameslamb** [comment]
rebased to master. If it builds tonight, I'm going to merge this. Thanks to everyone for the reviews!

**Comment by jameslamb** [comment]
😬 seems one task in azure pipelines has failed. Could someone with sufficient permissions re-run it? It's very very unlikely it could have been caused by this PR.

**Comment by StrikerRUS** [comment]
>😬 seems one task in azure pipelines has failed.

Seems that it was something connected to `apt-get`, maybe network issue.

```
Processing triggers for libc-bin (2.19-0ubuntu6.14) ...

bzip2: Compressed file ends unexpectedly;
	perhaps it is corrupted?  *Possible* reason follows.
bzip2: Inappropriate ioctl for device
	Input file = (stdin), output file = (stdout)

It is possible that the compressed file(s) have become corrupted.
You can use the -tvv option to test integrity of such files.

You can use the `bzip2recover' program to attempt to recover
data from undamaged sections of corrupted files.

tar: Child returned status 2
tar: Error is not recoverable: exiting now
sh: 0: Can't open AMD-APP-SDK*.sh
```

In addition, our `master` has failed too: https://lightgbm-ci.visualstudio.com/lightgbm-ci/_build/results?buildId=2380.

I have already spotted random failures in GPU task for several times and reported them in #1003. It seems that these bugs occure only in `test_cegb_scaling_equalities` from time to time.

**Comment by jameslamb** [comment]
Rebased to `master`. If this PR builds, I'll merge it tonight!

**Pull request merged**